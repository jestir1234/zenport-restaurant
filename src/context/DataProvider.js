import React, { Component } from "react";
import data from "../data.json";

export const DataContext = React.createContext();

const parseRestaurants = () => {
  let restaurants = {};
  let uniqRestaurants = Array.from(
    new Set(data[0].dishes.map(dish => dish.restaurant))
  );
  uniqRestaurants.forEach(name => {
    restaurants[name] = { availableMeals: "", menu: [] };
  });

  data[0].dishes.forEach(dish => {
    let currentMenu = restaurants[dish.restaurant].menu;
    currentMenu.push(dish.name);
    restaurants[dish.restaurant] = {
      availableMeals: dish.availableMeals,
      menu: currentMenu
    };
  });
  return restaurants;
};

const parseMealTypes = restaurants => {
  let mealTypes = {};
  Object.keys(restaurants).forEach(restaurant => {
    let restaurantMealTypes = restaurants[restaurant]["availableMeals"];
    restaurantMealTypes.forEach(type => {
      if (mealTypes[type]) {
        mealTypes[type]["restaurants"][restaurant] = restaurants[restaurant];
      } else {
        mealTypes[type] = { restaurants: {} };
      }
    });
  });
  return mealTypes;
};

const defaultStepOptions = () => ({
  step1: {
    selectedMealType: null,
    selectedPartyCount: { value: 1, label: 1 }
  },
  step2: {
    selectedRestaurant: null
  },
  step3: {
    selectedMeal: null,
    selectedServings: { value: 1, label: 1 },
    currentOrder: {},
    validMealCount: false,
    errorMessage: ""
  }
});

class DataProvider extends Component {
  state = {
    mealTypes: parseMealTypes(parseRestaurants()),
    stepOptions: defaultStepOptions()
  };

  handleChange = (selectedOption, option, currentStep) => {
    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        [currentStep]: {
          ...prevState.stepOptions[currentStep],
          [option]: selectedOption
        }
      }
    }));
  };

  addOrder = order => {
    let newOrder = { ...this.state.stepOptions.step3.currentOrder, ...order };
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder: newOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  subtractOrder = dish => {
    let currentOrder = { ...this.state.stepOptions.step3.currentOrder };
    delete currentOrder[dish];
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  validateMealCount = () => {
    let currentOrders = this.state.stepOptions.step3.currentOrder;
    let totalMeals =
      Object.keys(currentOrders).length > 0
        ? Object.values(this.state.stepOptions.step3.currentOrder).reduce(
            (acc, val) => acc + val
          )
        : 0;

    let selectedPartyCount = this.state.stepOptions.step1.selectedPartyCount
      .value;
    let errorMessage = "";

    if (totalMeals > 10) {
      errorMessage = "Exceeded max total servings of 10.";
    }

    if (totalMeals < selectedPartyCount) {
      errorMessage =
        "Total servings must be equal to or greater than the no. of people in your group.";
    }

    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        step3: {
          ...prevState.stepOptions.step3,
          validMealCount: totalMeals <= 10 && totalMeals >= selectedPartyCount,
          errorMessage
        }
      }
    }));
  };

  render() {
    return (
      <DataContext.Provider
        value={{
          state: this.state,
          handleChange: (selectedOption, option, currentStep) =>
            this.handleChange(selectedOption, option, currentStep),
          addOrder: order => this.addOrder(order),
          subtractOrder: dish => this.subtractOrder(dish)
        }}
      >
        {this.props.children}
      </DataContext.Provider>
    );
  }
}

export default DataProvider;
