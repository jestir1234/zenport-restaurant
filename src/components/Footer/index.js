import React from "react";
import styled, { css, keyframes } from "react-emotion";
import { DataContext } from "../../context/DataProvider";

const Container = styled("div")(props => ({
  width: "100%",
  height: "100px",
  border: "1px solid blue",
  display: "flex",
  justifyContent: "space-between",
  alignItems: "center"
}));

const buttonClass = css({
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  alignItems: "center",
  textAlign: "center",
  border: "1px solid #363c3c",
  fontWeight: "bold",
  letterSpacing: "2px",
  marginRight: "20px",
  cursor: "pointer",
  width: "150px",
  height: "30px",
  "&:hover": {
    background: "#b4fffd"
  },
  "&:active": {
    transform: "scale(.9)"
  }
});

const NextButton = styled("div")(props => ({
  marginRight: "20px",
  background: props.completed ? "#57dee1" : "#f0f0f0",
  pointerEvents: props.completed ? "auto" : "none"
}));

const PreviousButton = styled("div")(props => ({
  marginLeft: "20px",
  background: props.disableBack ? "#f0f0f0" : "#57dee1",
  pointerEvents: props.disableBack ? "none" : "auto"
}));

const stepCompleted = (currentPage, stepOptions) => {
  if (currentPage === "step4") {
    return true;
  }
  return !Object.values(stepOptions[currentPage]).includes(null);
};

const Footer = props => {
  return (
    <DataContext.Consumer>
      {context => {
        const completed = stepCompleted(
          props.currentPage.page,
          context.state.stepOptions
        );

        const disableBack = props.currentPage.page === "step1";
        return (
          <Container>
            <PreviousButton
              className={buttonClass}
              disableBack={disableBack}
              onClick={
                disableBack
                  ? null
                  : () => props.navigateToPreviousPage(props.currentPage)
              }
            >
              Previous
            </PreviousButton>
            <NextButton
              className={buttonClass}
              completed={completed}
              onClick={
                completed
                  ? () => props.navigateToNextPage(props.currentPage)
                  : null
              }
            >
              NEXT
            </NextButton>
          </Container>
        );
      }}
    </DataContext.Consumer>
  );
};

export default Footer;
