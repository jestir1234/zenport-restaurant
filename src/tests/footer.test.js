import React from "react";
import Footer from "../components/Footer";
import renderer from "react-test-renderer";
import DataProvider from "../context/DataProvider";

const currentPage = {
  next: "step2",
  page: "step1",
  prev: "step1"
};

const navigateToNextPage = () => {
  return true;
};

const navigateToPreviousPage = () => {
  return true;
};

test("Renders the <Footer /> component", () => {
  const component = renderer.create(
    <DataProvider>
      <Footer
        currentPage={currentPage}
        navigateToNextPage={navigateToNextPage}
        navigateToPreviousPage={navigateToPreviousPage}
      />
      }
    </DataProvider>
  );
  let tree = component.toJSON();
  expect(tree);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
