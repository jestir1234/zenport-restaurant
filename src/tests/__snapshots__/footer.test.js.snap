// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the <Footer /> component 1`] = `
Array [
  <div
    className="css-wxoicl"
>
    <div
        className="css-1ndri0n"
        onClick={null}
    >
        Previous
    </div>
    <div
        className="css-ve15fl"
        onClick={null}
    >
        Next
    </div>
</div>,
  "}",
]
`;
